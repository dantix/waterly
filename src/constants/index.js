export const STORAGE_KEY = 'waterly-data';
export const DATE_KEY_FORMAT = 'YYYY-MM-DD';
