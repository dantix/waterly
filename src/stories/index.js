import React from 'react';
import { storiesOf, action, linkTo } from '@kadira/storybook';

storiesOf('Header', module)
  .add('default', () => (
    <header>header</header>
  ));
