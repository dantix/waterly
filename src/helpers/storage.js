import { STORAGE_KEY } from '../constants';

export function get() {
  const data = localStorage.getItem(STORAGE_KEY);
  try {
    const result = JSON.parse(data);
    return result;
  } catch(e) {
    return {};
  }
}

export function set(data) {
  return localStorage.setItem(STORAGE_KEY, JSON.stringify(data));
}
