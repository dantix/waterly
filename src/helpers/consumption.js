import subDays from 'date-fns/sub_days';
import eachDay from 'date-fns/each_day';
import isAfter from 'date-fns/is_after';
import format from 'date-fns/format';

import { DATE_KEY_FORMAT } from '../constants';

import * as storage from './storage';

export function get({
  from = subDays(Date.now(), 2),
  to = Date.now(),
} = {}) {
  if (isAfter(from, to)) {
    throw new Error('"From" date should be before "to" date')
  }

  const data = storage.get();

  const range = eachDay(from, to);
  const keys = range.map(date => format(date, DATE_KEY_FORMAT))

  const result = keys.reduce((acc, key) => {
    acc[key] = data.consumption[key] || [];
    return acc;
  }, {});

  return result;
}

export function add({
  date = Date.now(),
  value,
} = {}) {
  if (!value) { throw new Error('No value provived'); }

  const data = storage.get();
  const key = format(date, DATE_KEY_FORMAT);

  const consumption = data.consumption[key] || [];
  consumption.push(value);

  data.consumption[key] = consumption;

  storage.set(data);
}

export function remove({ date = Date.now(), index } = {}) {
  if (typeof index === 'undefined') { throw new Error('No index provived'); }

  const data = storage.get();
  const key = format(date, DATE_KEY_FORMAT);

  const consumption = data.consumption[key] || [];
  const editedConsumption = [].concat(consumption.slice(0, index), consumption.slice(index + 1));
  data.consumption[key] = editedConsumption;

  storage.set(data);
}
