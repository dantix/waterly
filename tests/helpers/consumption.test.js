import format from 'date-fns/format';

import * as consumption from '../../src/helpers/consumption';

import { STORAGE_KEY, DATE_KEY_FORMAT } from '../../src/constants';

import { getLocalStorageStub } from '../__mocks__/localStorage';
import {
  initial,
  oneDay,
  defaultDateNow,
  defaultDateInPast
} from '../__fixtures__/consumption';

describe('consumption', () => {
  beforeAll(() => {
    global.localStorage = getLocalStorageStub();
  });

  afterEach(() => {
    global.localStorage.clear();
  });

  describe('get consumption', () => {
    it('should return data for 3 days if called without params', () => {
      global.localStorage.setFixture(oneDay);

      const entries = consumption.get({
        from: defaultDateInPast,
        to: defaultDateNow,
      });
      expect(entries).toMatchSnapshot();
    });

    it('should throw if "from" date is greater than "to"', () => {
      const assumption = () => consumption.get({
        from: new Date(2017, 3, 24),
        to: new Date(2017, 3, 22),
      });

      expect(assumption).toThrowErrorMatchingSnapshot();
    });
  });

  describe('add consumption', () => {
    it('should add entry', () => {
      global.localStorage.setFixture(oneDay);

      const testEntry = {
        type: 'tea',
        amount: 0.5,
      };

      consumption.add({ value: testEntry });

      const todayKey = format(defaultDateNow, DATE_KEY_FORMAT);

      const str = global.localStorage.getItem(STORAGE_KEY);
      const data = JSON.parse(str);

      const consumptionEntries = data.consumption[todayKey];
      const lastEntry = consumptionEntries[consumptionEntries.length - 1];

      expect(lastEntry).toEqual(testEntry);
    });

    it('should throw if no value provided', () => {
      const assumption = () => consumption.add();
      expect(assumption).toThrowErrorMatchingSnapshot();
    });
  });

  describe('remove consumption', () => {
    it('should remove entry', () => {
      global.localStorage.setFixture(oneDay);

      consumption.remove({
        date: defaultDateNow,
        index: 0,
      });

      const str = global.localStorage.getItem(STORAGE_KEY);
      const data = JSON.parse(str);

      const todayKey = format(defaultDateNow, DATE_KEY_FORMAT);

      const consumptionEntries = data.consumption[todayKey];
      expect(consumptionEntries).toHaveLength(1);
      expect(consumptionEntries).toMatchSnapshot();
    });

    it('should throw if no index provided', () => {
      const assumption = () => consumption.remove();
      expect(assumption).toThrowErrorMatchingSnapshot();
    });
  });
})
