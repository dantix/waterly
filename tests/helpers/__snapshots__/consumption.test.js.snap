exports[`consumption add consumption should throw if no value provided 1`] = `"No value provived"`;

exports[`consumption get consumption should return data for 3 days if called without params 1`] = `
Object {
  "2017-04-22": Array [],
  "2017-04-23": Array [],
  "2017-04-24": Array [
    Object {
      "amount": 0.3,
      "type": "water",
    },
    Object {
      "amount": 0.5,
      "type": "tea",
    },
  ],
}
`;

exports[`consumption get consumption should throw if "from" date is greater than "to" 1`] = `"\"From\" date should be before \"to\" date"`;

exports[`consumption remove consumption should remove entry 1`] = `
Array [
  Object {
    "amount": 0.5,
    "type": "tea",
  },
]
`;

exports[`consumption remove consumption should throw if no index provided 1`] = `"No index provived"`;
